azure_gateway/_gateway_llm.py

```diff
@@ -0,0 +1,282 @@
+import json
+import logging
+import os
+import time
+from typing import Callable
+
+import requests
+import tiktoken
+from langchain_core.language_models.llms import LLM
+from openai import AzureOpenAI
+from requests_sse import EventSource
+
+from ._exceptions import OutOfQuotaError
+from ._logger import LoggerWrapper
+
+
+Callback = Callable[[str], None]
+ServiceResponse = tuple[str, int, int]
+
+
+class BaseGatewayLLM(LLM):
+    """Base class for Gateway LLM"""
+    project_id: str
+    model_id: str
+    model_type: str
+    conversation_history: list
+    gateway_auth_token: str
+    alert_log_threshold: float = 0.9
+    logger: logging.Logger | None = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.log = LoggerWrapper(self.logger)
+
+    @property
+    def _llm_type(self) -> str:
+        return "custom"
+
+    @property
+    def _identifying_params(self):
+        """Get the identifying parameters."""
+        return {}
+
+    def _call(self, prompt: str, callback: Callback | None, **_):
+        self._query_gateway_safely(query_function=self._check_quota)
+
+        self.log.info(f"PROMPT\n{self.log.indent_text(prompt)}")
+        self.conversation_history.append({"role": "user", "content": prompt})
+
+        response_text, prompt_tokens, completion_tokens = self._call_service(callback=callback)
+
+        self._query_gateway_safely(
+            query_function=lambda: self._update_quota(
```

Comment:
I'm not sure I understand the issue. `_query_gateway_safely()` should do this: re-raise `OutOfQuotaError` in case that's the original error or just log any other error and move on. It ensures a project never stops working because of a random error, but it should stop when the quota is full. What would be the benefit of not ignoring errors in this case?

Note: I didn't implement this so this is only my opinion on why it might be this way.